ipam/allocator.go

```diff
@@ -159,389 +211,414 @@ const (
+		a.updateRefCount(p, 1)
+		return func() error { return nil }, nil
 	}
-	return false
-}
-
-// Splits the passed subnet into N internal subnets with host size equal to internalHostSize.
-// If the subnet's host size is equal to or smaller than internalHostSize, there won't be any
-// split and the return list will contain only the passed subnet.
-func getInternalSubnets(inSubnet *net.IPNet, internalHostSize int) ([]*net.IPNet, error) {
-	var subnetList []*net.IPNet
 
-	// Sanity check and size adjustment for v6
-	subnet, err := adjustAndCheckSubnetSize(inSubnet)
-	if err != nil {
-		return subnetList, err
+	// If master pool, check for overlap
+	if ipr == nil {
+		if a.contains(k.AddressSpace, nw) {
+			return nil, ipamapi.ErrPoolOverlap
+		}
+		// This is a new master pool, add it along with corresponding bitmask
+		a.Lock()
+		a.subnets[k] = &PoolData{Pool: nw, RefCount: 1}
+		a.Unlock()
+		return func() error { return a.insertBitMask(k, nw) }, nil
 	}
 
-	// Get network/host subnet information
-	netBits, bits := subnet.Mask.Size()
-	hostBits := bits - netBits
+	// This is a new non-master pool
+	p = &PoolData{
+		ParentKey: SubnetKey{AddressSpace: k.AddressSpace, Subnet: k.Subnet},
+		Pool:      nw,
+		Range:     ipr,
+		RefCount:  1,
+	}
+	a.Lock()
+	a.subnets[k] = p
+	a.Unlock()
 
-	extraBits := hostBits - internalHostSize
-	if extraBits <= 0 {
-		subnetList = make([]*net.IPNet, 1)
-		subnetList[0] = subnet
-	} else {
-		// Split in smaller internal subnets
-		numIntSubs := 1 << uint(extraBits)
-		subnetList = make([]*net.IPNet, numIntSubs)
-
-		// Construct one copy of the internal subnets's mask
-		intNetBits := bits - internalHostSize
-		intMask := net.CIDRMask(intNetBits, bits)
-
-		// Construct the prefix portion for each internal subnet
-		for i := 0; i < numIntSubs; i++ {
-			intIP := make([]byte, len(subnet.IP))
-			copy(intIP, subnet.IP) // IPv6 is too big, just work on the extra portion
-			addIntToIP(intIP, uint32(i<<uint(internalHostSize)))
-			subnetList[i] = &net.IPNet{IP: intIP, Mask: intMask}
-		}
+	// Look for parent pool
+	a.Lock()
+	pp, ok := a.subnets[p.ParentKey]
+	a.Unlock()
+	if ok {
+		a.updateRefCount(pp, 1)
+		return func() error { return nil }, nil
 	}
-	return subnetList, nil
+	// Parent pool does not exist, add it along with corresponding bitmask
+	a.Lock()
+	a.subnets[p.ParentKey] = &PoolData{Pool: nw, RefCount: 1}
+	a.Unlock()
+	return func() error { return a.insertBitMask(p.ParentKey, nw) }, nil
 }
 
-// RemoveSubnet removes the subnet from the specified address space
-func (a *Allocator) RemoveSubnet(addrSpace AddressSpace, subnet *net.IPNet) error {
-	if addrSpace == "" {
-		return ErrInvalidAddressSpace
-	}
-	if subnet == nil {
-		return ErrInvalidSubnet
-	}
-retry:
-	// Look for the respective subnet configuration data
-	// Remove it along with the internal subnets
-	subKey := subnetKey{addrSpace, subnet.String(), ""}
+func (a *Allocator) updatePoolDBOnRemoval(k SubnetKey) (func() error, error) {
```

Comment:
Let's keep this whole function under lock because there is a race when you are traversing the tree while a `updatePoolDBonAdd` is happening in parallel